### ISSUE TITLE
Redesign database structure & refactor shop system

### ISSUE THREAD

**Issue by Ghost-chu** [Ghost-chu/QuickShop-Hikari#179]
## Why
The current database structure design and store system design are based on block coordinates.  
This means that all the stores' data stores are binded to the coordinates of the Block.

That's mean, GUI shop and multi-shops in one block is impossible to do with the current data structure.

## Goal

* Make the store's metadata not strongly associated with Block coordinates.
* Modify data structure to be compatible with GUI shop, multi-shop etc.
* Restructured plugin structure, modularization of QuickShop features.
  * This should help advance the de-Playerization and use UUIDs in the internal core to do all the work.
    * In other words, it doesn't have to be a physically present player to trade with the store, which should make it easier for other Addons to simulate transactions while a particular player is offline.
* Cross server data sync in real-time. (Feat. [MultiPaper](https://github.com/PureGero/MultiPaper))

**Comment by kohle** [comment]
Hope it's okay to throw my two cents in here!

I'm interested in using the data that QuickShop logs to perform an analysis on the overall server economy. Thanks for implementing MySQL, that makes it a lot easier to run queries 😁

**qs_metrics**

- Is this supposed to be an overall log of all shop activities?
- When I remove a shop, the `type` is still recorded as `CREATE` but the price is set to `0.00`
  - This _might_ be okay except that I'm allowing free shops... `DELETE` or something along those lines would be much better!
- Unfortunately, removing a shop also removes it from `qs_shops`. So even though the transaction (thank you for recording these) is kept, there's no way to see _which_ shop the transaction took place with.
  - Would it make sense to either set a boolean like `active` to mark if a shop is deleted or not in `qs_shops` and have a foreign key in the `qs_metrics` table? Sounds like overall you want to move more towards indexing each shop rather than basing it on the coordinates (100% think this would be much better!)

![image](https://user-images.githubusercontent.com/892288/168503371-26d7640f-7f57-4a15-8869-870e7d619e2c.png)

Essentially, it would be amazing if the database:

- Keeps track of all economic transactions
- Records the history of chest shops, so transactions can be matched to the item, quantity, and players (owner and customer)

I absolutely love the usability of this plugin, so a better database structure would be terrific! Think about how cool it would be to have a web dashboard where you can see things like all active shops, the items they are selling, and even historical price data! Instead of having players make random prices, you could query the database to see the average transaction price of any item. So many cool things to do with data 😄

**Comment by YuanYuanOwO** [comment]
> Hope it's okay to throw my two cents in here!
> 
> I'm interested in using the data that QuickShop logs to perform an analysis on the overall server economy. Thanks for implementing MySQL, that makes it a lot easier to run queries 😁
> 
> **qs_metrics**
> 
> * Is this supposed to be an overall log of all shop activities?
> * When I remove a shop, the `type` is still recorded as `CREATE` but the price is set to `0.00`
>   
>   * This _might_ be okay except that I'm allowing free shops... `DELETE` or something along those lines would be much better!
> * Unfortunately, removing a shop also removes it from `qs_shops`. So even though the transaction (thank you for recording these) is kept, there's no way to see _which_ shop the transaction took place with.
>   
>   * Would it make sense to either set a boolean like `active` to mark if a shop is deleted or not in `qs_shops` and have a foreign key in the `qs_metrics` table? Sounds like overall you want to move more towards indexing each shop rather than basing it on the coordinates (100% think this would be much better!)
> 
> ![image](https://user-images.githubusercontent.com/892288/168503371-26d7640f-7f57-4a15-8869-870e7d619e2c.png)
> 
> Essentially, it would be amazing if the database:
> 
> * Keeps track of all economic transactions
> * Records the history of chest shops, so transactions can be matched to the item, quantity, and players (owner and customer)
> 
> I absolutely love the usability of this plugin, so a better database structure would be terrific! Think about how cool it would be to have a web dashboard where you can see things like all active shops, the items they are selling, and even historical price data! Instead of having players make random prices, you could query the database to see the average transaction price of any item. So many cool things to do with data 😄

https://github.com/Ghost-chu/QuickShop-Hikari/wiki/Database-structure  
hope this can help you

**Comment by CarmJos** [comment]
Captain, my captain, I am on standby.

**Comment by Ghost-chu** [comment]
> Captain, my captain, I am on standby.

Glad to see you have joined database redesign project.
Write access has been granted.

**Comment by CarmJos** [comment]
Please check my commit at https://github.com/Ghost-chu/QuickShop-Hikari/commit/def79379da6f6bb4c5388e8de1ed79c6d82a85ba

**Comment by CarmJos** [comment]
My job is done, bye.
